File: getting-started/variant-ts/module-08-modules-and-composition/exercise-02-deep-vs-shallow/stub/option_a/cache.ts
// Option A: The shallow approach
//
// Every internal detail is exposed. The caller manages expiration,
// creates entries, checks timestamps, and drives the cleanup.

export type CacheEntry = {
  key: string;
  value: string;
  createdAt: number;  // Date.now() timestamp
  expiresAt: number;  // Date.now() timestamp
  isExpired: boolean;
};

export class Cache {
  entries: Map<string, CacheEntry> = new Map();

  putEntry(entry: CacheEntry): void {
    this.entries.set(entry.key, entry);
  }

  getEntry(key: string): CacheEntry | undefined {
    return this.entries.get(key);
  }

  deleteEntry(key: string): void {
    this.entries.delete(key);
  }

  cleanupExpired(): number {
    let removed = 0;
    for (const [key, entry] of this.entries) {
      if (checkExpired(entry)) {
        this.entries.delete(key);
        removed++;
      }
    }
    return removed;
  }

  size(): number {
    return this.entries.size;
  }
}

export const createEntry = (key: string, value: string, ttlMs: number): CacheEntry => {
  const now = Date.now();
  return {
    key,
    value,
    createdAt: now,
    expiresAt: now + ttlMs,
    isExpired: false,
  };
};

export const checkExpired = (entry: CacheEntry): boolean =>
  Date.now() > entry.expiresAt;

File: getting-started/variant-go/module-08-modules-and-composition/exercise-02-deep-vs-shallow/stub/option_a/cache.go
// Option A: The shallow approach
//
// Every internal detail is exposed. The caller manages expiration,
// creates entries, checks timestamps, and drives the cleanup.

package option_a

import "time"

type CacheEntry struct {
	Key       string
	Value     string
	CreatedAt time.Time
	ExpiresAt time.Time
	IsExpired bool
}

type Cache struct {
	Entries map[string]CacheEntry
}

func NewCache() *Cache {
	return &Cache{Entries: make(map[string]CacheEntry)}
}

func CreateEntry(key, value string, ttl time.Duration) CacheEntry {
	now := time.Now()
	return CacheEntry{
		Key:       key,
		Value:     value,
		CreatedAt: now,
		ExpiresAt: now.Add(ttl),
		IsExpired: false,
	}
}

func (c *Cache) PutEntry(entry CacheEntry) {
	c.Entries[entry.Key] = entry
}

func (c *Cache) GetEntry(key string) (CacheEntry, bool) {
	entry, ok := c.Entries[key]
	return entry, ok
}

func (c *Cache) DeleteEntry(key string) {
	delete(c.Entries, key)
}

func CheckExpired(entry CacheEntry) bool {
	return time.Now().After(entry.ExpiresAt)
}

func (c *Cache) CleanupExpired() int {
	removed := 0
	for key, entry := range c.Entries {
		if CheckExpired(entry) {
			delete(c.Entries, key)
			removed++
		}
	}
	return removed
}

func (c *Cache) Size() int {
	return len(c.Entries)
}
